sklearn/tests/test_calibration.py

```diff
@@ -338,42 +388,56 @@ def test_calibration_nan_imputer():
     clf = Pipeline(
         [('imputer', SimpleImputer()),
          ('rf', RandomForestClassifier(n_estimators=1))])
-    clf_c = CalibratedClassifierCV(clf, cv=2, method='isotonic')
+    clf_c = CalibratedClassifierCV(
+        clf, cv=2, method='isotonic', ensemble=ensemble
+    )
     clf_c.fit(X, y)
     clf_c.predict(X)
 
 
-def test_calibration_prob_sum():
+@pytest.mark.parametrize('ensemble', [True, False])
+def test_calibration_prob_sum(ensemble):
     # Test that sum of probabilities is 1. A non-regression test for
     # issue #7796
     num_classes = 2
     X, y = make_classification(n_samples=10, n_features=5,
                                n_classes=num_classes)
-    clf = LinearSVC(C=1.0)
-    clf_prob = CalibratedClassifierCV(clf, method="sigmoid", cv=LeaveOneOut())
+    clf = LinearSVC(C=1.0, random_state=7)
+    clf_prob = CalibratedClassifierCV(
+        clf, method="sigmoid", cv=LeaveOneOut(), ensemble=ensemble
+    )
     clf_prob.fit(X, y)
 
     probs = clf_prob.predict_proba(X)
     assert_array_almost_equal(probs.sum(axis=1), np.ones(probs.shape[0]))
 
 
-def test_calibration_less_classes():
+@pytest.mark.parametrize('ensemble', [True, False])
+def test_calibration_less_classes(ensemble):
     # Test to check calibration works fine when train set in a test-train
     # split does not contain all classes
     # Since this test uses LOO, at each iteration train set will not contain a
     # class label
     X = np.random.randn(10, 5)
     y = np.arange(10)
-    clf = LinearSVC(C=1.0)
-    cal_clf = CalibratedClassifierCV(clf, method="sigmoid", cv=LeaveOneOut())
+    clf = LinearSVC(C=1.0, random_state=7)
+    cal_clf = CalibratedClassifierCV(
+        clf, method="sigmoid", cv=LeaveOneOut(), ensemble=ensemble
+    )
     cal_clf.fit(X, y)
 
     for i, calibrated_classifier in \
             enumerate(cal_clf.calibrated_classifiers_):
         proba = calibrated_classifier.predict_proba(X)
-        assert_array_equal(proba[:, i], np.zeros(len(y)))
-        assert np.all(np.hstack([proba[:, :i],
-                                 proba[:, i + 1:]]))
+        if ensemble:
+            # Check 'missing' class has proba=0
+            assert_array_equal(proba[:, i], np.zeros(len(y)))
+            # Check for all other classes proba!=0
+            assert np.all(np.hstack([proba[:, :i],
+                                     proba[:, i + 1:]]))
+        else:
+            # Check `proba` all 1/n_classes
```

Comment:
am I correct to assume this is because the regressor was trained on an array with only zeros as features?

Reply:
I think it's because each sample is its own class - there are 10 samples and 10 classes. With `LeaveOneOut` cross-validation, the test set contains just one sample, and this sample belongs to a class not present in the training subset! So the prediction output is always the same value for every test subset (for `LinearSVC` at least).

I tried `cross_val_predict` with this data locally and I actually get the warning, which I guess is ignored when we use pytest (?):

> RuntimeWarning: Number of classes in training fold (9) does not match total number of classes (10). Results may not be appropriate for your use case. To fix this, use a cross-validation technique resulting in properly stratified folds

It's a bit odd, but I think each sample has to be it's own class for us to be able to test the results easily. Though this is confusing and I am not certain.

FYI It's related to this issue #7796 (fixed in #7799), where the probas weren't summing to 1 with `LeaveOneOut` because the test subset (1 sample) contained less classes than the train subset.